common/src/address.rs

```diff
@@ -177,6 +177,256 @@ pub fn get_64_subnet(
+#[serde(try_from = "AnyIpv4Range")]
+pub struct Ipv4Range {
+    pub first: Ipv4Addr,
+    pub last: Ipv4Addr,
+}
+
+impl Ipv4Range {
+    pub fn new(first: Ipv4Addr, last: Ipv4Addr) -> Result<Self, String> {
+        if first <= last {
+            Ok(Self { first, last })
+        } else {
+            Err(String::from("IP address ranges must be non-decreasing"))
+        }
+    }
+
+    pub fn first_address(&self) -> Ipv4Addr {
+        self.first
+    }
+
+    pub fn last_address(&self) -> Ipv4Addr {
+        self.last
+    }
+
+    pub fn iter(&self) -> Ipv4RangeIter {
+        Ipv4RangeIter { next: Some(self.first.into()), last: self.last.into() }
+    }
+}
+
+impl From<Ipv4Addr> for Ipv4Range {
+    fn from(addr: Ipv4Addr) -> Self {
+        Self { first: addr, last: addr }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+struct AnyIpv4Range {
+    first: Ipv4Addr,
+    last: Ipv4Addr,
+}
+
+impl TryFrom<AnyIpv4Range> for Ipv4Range {
+    type Error = Error;
+    fn try_from(r: AnyIpv4Range) -> Result<Self, Self::Error> {
+        Ipv4Range::new(r.first, r.last)
+            .map_err(|msg| Error::invalid_request(msg.as_str()))
+    }
+}
+
+/// A non-decreasing IPv6 address range, inclusive of both ends.
+///
+/// The first address must be less than or equal to the last address.
+#[derive(Clone, Copy, Debug, PartialEq, Deserialize, Serialize, JsonSchema)]
+#[serde(try_from = "AnyIpv6Range")]
+pub struct Ipv6Range {
+    pub first: Ipv6Addr,
+    pub last: Ipv6Addr,
+}
+
+impl Ipv6Range {
+    pub fn new(first: Ipv6Addr, last: Ipv6Addr) -> Result<Self, String> {
+        if first <= last {
+            Ok(Self { first, last })
+        } else {
+            Err(String::from("IP address ranges must be non-decreasing"))
+        }
+    }
+
+    pub fn first_address(&self) -> Ipv6Addr {
+        self.first
+    }
+
+    pub fn last_address(&self) -> Ipv6Addr {
+        self.last
+    }
+
+    pub fn iter(&self) -> Ipv6RangeIter {
+        Ipv6RangeIter { next: Some(self.first.into()), last: self.last.into() }
+    }
+}
+
+impl From<Ipv6Addr> for Ipv6Range {
+    fn from(addr: Ipv6Addr) -> Self {
+        Self { first: addr, last: addr }
+    }
+}
+
+#[derive(Clone, Copy, Debug, Deserialize)]
+struct AnyIpv6Range {
+    first: Ipv6Addr,
+    last: Ipv6Addr,
+}
+
+impl TryFrom<AnyIpv6Range> for Ipv6Range {
+    type Error = Error;
+    fn try_from(r: AnyIpv6Range) -> Result<Self, Self::Error> {
+        Ipv6Range::new(r.first, r.last)
+            .map_err(|msg| Error::invalid_request(msg.as_str()))
+    }
+}
+
+pub struct Ipv4RangeIter {
+    next: Option<u32>,
+    last: u32,
+}
+
+impl Iterator for Ipv4RangeIter {
+    type Item = Ipv4Addr;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.next?;
+        if next < self.last {
+            self.next = Some(next + 1);
+        } else {
+            self.next = None;
+        }
+        Some(next.into())
+    }
+}
```

Comment:
This could probably be a generic over u32 vs u128 but also frankly if you wanna keep 'em separate I get it